os/20201103-full-disk-raid.md

```diff
@@ -0,0 +1,208 @@
+
+`/boot` replication prevents the use of the grubenv mechanism.  If GRUB treats the replicas as independent filesystems, grubenv will cause the RAID to desynchronize.  Alternatively, if we were to use the GRUB RAID module, GRUB would [disable grubenv support](https://www.gnu.org/software/grub/manual/grub/grub.html#Environment-block).
+
+## EFI System Partition
+
+The Dracut glue will save the contents of the ESP filesystem whenever the config specifies a partition with the ESP type GUID, and will restore those contents to _every_ partition with that type GUID.
+
+Thus the ESP won't be RAIDed; we'll create multiple independent filesystems with identical directory trees.  Keeping each replica independent avoids RAID desynchronization if the firmware chooses to modify the ESP, and makes sense because CoreOS does not modify the ESP after installation.  Since there will no longer be a single canonical ESP, we'll no longer automount `/boot/efi` in the running system.
+
+When updating the bootloader, bootupd will need to find each partition with an ESP type GUID and update it independently.  This assumes that all ESPs on the system are controlled by CoreOS.
+
+## BIOS Boot partition
+
+The Dracut glue will save an image of the BIOS boot partition whenever the config specifies a partition with the BIOS boot type GUID, and will restore that image to _every_ such partition.  In addition, the same logic will save the boot sector (the first 440 bytes of the disk) and restore it to every disk where we write a BIOS Boot partition.
+
+The BIOS Boot partition is also not updated at runtime, so RAID is not necessary here either.  As a result, when updating the bootloader, bootupd will need to find each BIOS Boot type GUID and update both the partition and the corresponding boot sector.  Again this assumes that the entire system belongs to CoreOS.
+
+For simplicity, we're copying the BIOS Boot partition and boot sector bit-for-bit, rather than rerunning `grub-install` from the initrd or hand-modifying the boot sector.  Since the boot sector hardcodes the offset of the BIOS Boot partition, the latter must be recreated at the same offset as the original partition.  To avoid hardcoding an awkward constant in FCCT, we'll move the BIOS boot partition from its current offset of 512 MiB to the beginning of the disk (offset 1 MiB) in new images.  The Dracut glue will detect relocation of the BIOS boot partition and fail the boot.
+
+There's no BIOS boot partition in 4Kn images.  We don't know in advance which disks will be 4Kn, so the Ignition config will always create a BIOS boot partition on x86_64, but the Dracut glue will automatically skip copying it (and the boot sector) if there's no source partition to copy from.
```

Comment:
@jlebon and I talked in the past and we seemed to agree that for simplicity sake it would be nice if the partition scheme even for 4k disks matched the non-4k disks. i.e. we were proposing at the time that we have the same number of partitions for both types even if it's not used for 4k.

WDYT?

Reply:
If folks are up for it, yeah, I'd prefer that we reduce the number of distinct partition layouts.  :+1: